engine/consensus/matching/engine.go

```diff
@@ -1168,46 +1188,72 @@ func (e *Engine) requestPendingReceipts() (int, uint64, error) {
-		// get finalized, so the returned receipts must be from finalized blocks.
-		// Therefore, the return receipts must be incoporated receipts, which
-		// are safe to be added to the mempool
+		// Without the logic below, the matching engine would produce IncorporatedResults
+		// only from receipts received directly from ENs. Matching Core would not know about
+		// Receipts that are incorporated by other nodes in their blocks blocks (but never
+		// received directly from the EN). Also, receipts might have been lost from the
+		// mempool during a node crash. Hence we check also if we have the receipts in
+		// storage (which also persists receipts pre-crash or when received from other
+		// nodes as part of a block proposal).
+		// Currently, receipts are only indexed by the block which they compute when:
+		// (a) A block containing the receipt is incorporated into the main chain.
+		//     Therefore, the receipt must have passed validation
+		// (b) The matching engine processes and validates the receipt
+		// In either case, only valid receipts are index.
+		// CAUTION: shortcut for Sealing Phase 2:
+		// * The matching engine adds any known receipts for unsealed blocks
+		//   directly to the IncorporatedResults mempool. Hence, we do the same
+		//   for any receipts from the storage layer.
+		// ToDo: this logic must be updated for Sealing Phase 3
+		// OnBlockIncorporated callback planned for phase 3 of the S&V roadmap,
+		// and that the IncorporatedResult's IncorporatedBlockID should be set
+		// correctly.
 		receipts, err := e.receiptsDB.ByBlockIDAllExecutionReceipts(blockID)
 		if err != nil {
 			return 0, 0, fmt.Errorf("could not get receipts by block ID: %v, %w", blockID, err)
 		}
 
-		if len(receipts) > 0 {
-			for _, receipt := range receipts {
-				_, err = e.receipts.AddReceipt(receipt, header)
-				if err != nil {
-					return 0, 0, fmt.Errorf("could not add receipt to receipts mempool %v, %w", receipt.ID(), err)
-				}
+		enoughReceipts := false
+		receiptsForResult := make(map[flow.Identifier]int) // map: resultID -> number of receipts committing to this result
+		for _, receipt := range receipts {
 
-				_, err = e.incorporatedResults.Add(
-					flow.NewIncorporatedResult(
-						receipt.ExecutionResult.BlockID,
-						&receipt.ExecutionResult,
-					),
-				)
+			_, err = e.receipts.AddReceipt(receipt, header)
+			if err != nil {
+				return 0, 0, fmt.Errorf("could not add receipt to receipts mempool %v, %w", receipt.ID(), err)
+			}
 
-				if err != nil {
-					return 0, 0, fmt.Errorf("could not add result to incorporated results mempool %v, %w", receipt.ID(), err)
-				}
+			_, err = e.incorporatedResults.Add(
+				flow.NewIncorporatedResult(receipt.ExecutionResult.BlockID, &receipt.ExecutionResult),
+			)
+			if err != nil {
+				return 0, 0, fmt.Errorf("could not add result to incorporated results mempool %v, %w", receipt.ID(), err)
 			}
+
```

Comment:
```suggestion
			if enoughReceipts {
				continue
			}
```

Reply:
Do we need to update `receiptsForResult` if we had enough receipts?

Reply:
Yea, we could shortcut it. My thinking is:
* The expensive operations are probably the mempool additions. We could not add receipts to mempools if we already have a sufficient number. My thinking was that it was fine to accept the extra cost for the benefit of keeping mempool and storage in sync. 
* But if we are happy to eat the extra cost for mempool additions, I don't think the few extra operations for incrementing the counter would make any difference.